aacplusenc.1
.TH "AACPLUSENC" "1" "19 January 2008" "Matteo Croce" "Debian GNU/Linux"
.SH "NAME"
aacplusenc \- AAC+ encoder
.SH "SYNOPSIS"
.B aacplusenc
.I inputfile.wav outputfile.aac bitrate
.PP
.SH "DESCRIPTION"
\fBaacplusenc\fP is an \fIHE\-AACv2\fP encoder.

It accepts as input a \fIWAV\fP file and outputs an \fIAAC\fP file, with \fIRAW ADTS\fP header.
.PP
.SH "OPTIONS"
.TP 8
.B inputfile.wav
The file to encode
.TP 8
.B outputfile.aac
The desired filename of the encoded file
.TP 8
.B bitrate
The encoding bitrate, in kbps
.TP 8
both inputfile and outputfile can be \fI"\-"\fP to specify stdin and stdout respectively
.SH "NOTES"
On 64 bit platform there are issues with Parametric Stereo (PS) an algorithm to give
better quality on stereo tracks.
.PP
As PS is enabled for bitrates lower than 45 kbps, a 64 bit build of \fBaacplusenc\fP can only encode
mono tracks or stereo tracks encoded with bitrates of 45 kbps and over
.SH "AUTHOR"
\fBaacplusenc\fP is written by Matteo Croce <\fIrootkit85@yahoo.it\fP> and its based on 3GPP sources.

Get the latest version from \fIhttps://github.com/teknoraver/aacplusenc\fP

The 3GPP project used by \fBaacplusenc\fP: for more info visit \fIhttp://www.3gpp.org/ftp/Specs/archive/26_series/26.410/\fP

AAC+ is licensed by Coding Technologies, for info read: \fIhttp://codtech.vhost.noris.net/products/aacPlus.htm\fP

This manpage was written by Matteo Croce <\fIrootkit85@yahoo.it\fP> and licensed under the terms of the GNU General Public License.


debian/changelog
aacplusenc (0.17.4) jaunty; urgency=low

  * Fix ppc encoding
  * Jaunty upload

 -- Matteo Croce <rootkit85@yahoo.it>  Mon, 18 May 2009 18:08:17 +0200

aacplusenc (0.17.3) intrepid; urgency=low

  * Fix possibly corrupted ADTS header
  * Intrepid upload

 -- Matteo Croce <rootkit85@yahoo.it>  Wed, 29 Oct 2008 23:50:12 +0100

aacplusenc (0.17.2) intrepid; urgency=low

  * Intrepid upload

 -- Matteo Croce <rootkit85@yahoo.it>  Sat, 13 Sep 2008 01:22:11 +0200

aacplusenc (0.17.1) hardy; urgency=low

  * Updated homepage URL

 -- Matteo Croce <rootkit85@yahoo.it>  Thu, 17 Jul 2008 16:44:46 +0200

aacplusenc (0.17) hardy; urgency=high

  * Fixed 64 bit encoding

 -- Matteo Croce <rootkit85@yahoo.it>  Wed, 04 Jun 2008 00:42:44 +0200

aacplusenc (0.16.3) hardy; urgency=low

  * Fix type in au_channel.h

 -- Matteo Croce <rootkit85@yahoo.it>  Sun, 17 Feb 2008 15:55:51 +0100

aacplusenc (0.16.2) hardy; urgency=low

  * Added compile time checks for endiannes and FFTW3

 -- Matteo Croce <rootkit85@yahoo.it>  Thu, 14 Feb 2008 16:32:42 +0100

aacplusenc (0.16.1) hardy; urgency=low

  * Fix BSD build

 -- Matteo Croce <rootkit85@yahoo.it>  Tue, 12 Feb 2008 17:39:16 +0100

aacplusenc (0.16) hardy; urgency=low

  * Added support for bitrate up to 64 and 63 kbit for 44.1 and 48 kHz

 -- Matteo Croce <rootkit85@yahoo.it>  Thu, 31 Jan 2008 18:48:17 +0100

aacplusenc (0.15) hardy; urgency=low

  * Full 64 bit support

 -- Matteo Croce <rootkit85@yahoo.it>  Wed, 23 Jan 2008 20:14:58 +0100

aacplusenc (0.14.2) hardy; urgency=low

  * Rebuild ubuntu package with sane dpkg-deb arguments

 -- Matteo Croce <rootkit85@yahoo.it>  Wed, 23 Jan 2008 16:01:31 +0100

aacplusenc (0.14.1) hardy; urgency=low

  * Fix endiannes under BSD

 -- Matteo Croce <rootkit85@yahoo.it>  Wed, 23 Jan 2008 01:20:42 +0100

aacplusenc (0.14) hardy; urgency=low

  * Small speedup (~9%)
  * Better packaging

 -- Matteo Croce <rootkit85@yahoo.it>  Tue, 22 Jan 2008 03:26:28 +0100

aacplusenc (0.13.2) hardy; urgency=low

  * Fix manpage path

 -- Matteo Croce <rootkit85@yahoo.it>  Sun, 20 Jan 2008 04:18:55 +0100

aacplusenc (0.13.1-1~gutsy1) gutsy; urgency=low

  * Gutsy upload

 -- Matteo Croce <rootkit85@yahoo.it>  Sun, 20 Jan 2008 00:15:31 +0100

aacplusenc (0.13.1) hardy; urgency=low

  * Added changelog

 -- Matteo Croce <rootkit85@yahoo.it>  Sat, 19 Jan 2008 23:56:44 +0100

aacplusenc (0.13) gutsy; urgency=low

  * All channels/samplerates/bitrates combinations works now
  * A 64 bit build can encode a mono track, or an high bitrate one (> 44kbit)

 -- Matteo Croce <rootkit85@yahoo.it>  Tue, 15 Jan 2008 01:23:42 +0100

aacplusenc (0.12.1) gutsy; urgency=low

  * Fix Win32 build

 -- Matteo Croce <rootkit85@yahoo.it>  Mon, 17 Dec 2007 17:36:39 +0100

aacplusenc (0.12) gutsy; urgency=low

  * Fixed encodings higher tha 44 kbit/s

 -- Matteo Croce <rootkit85@yahoo.it>  Thu, 06 Dec 2007 02:16:07 +0100

aacplusenc (0.11) gutsy; urgency=low

  * Enable 64 bit builds, but only for mono encodings

 -- Matteo Croce <rootkit85@yahoo.it>  Tue, 04 Dec 2007 17:53:59 +0100

aacplusenc (0.10) gutsy; urgency=low

  * Good speed increase (25-33%) against older version
  * Keep legacy builtin FFT, enable it with the NOFFTW3 make flag

 -- Matteo Croce <rootkit85@yahoo.it>  Sat, 01 Dec 2007 18:10:30 +0100

aacplusenc (0.9) gutsy; urgency=low

  * Trashed the builtin FFT and used the GPL one FFTW3 which is faster

 -- Matteo Croce <rootkit85@yahoo.it>  Fri, 30 Nov 2007 02:16:58 +0100

aacplusenc (0.8) gutsy; urgency=low

  * removed the last m/s parameter. Stereo or mono encoding
    is done according to source number of channels
  * now the target bitrate is expressed in kbit, not bits

 -- Matteo Croce <rootkit85@yahoo.it>  Tue, 20 Nov 2007 18:35:37 +0100

aacplusenc (0.7) gutsy; urgency=low

  * Adding safety code to avoid encoder runtime-assertion

 -- Matteo Croce <rootkit85@yahoo.it>  Fri, 19 Oct 2007 02:09:05 +0200

aacplusenc (0.6) feisty; urgency=low

  * fixed percentage display
  * copied missing b2l macro from old libaudio/au_channel.h
  * removed unneeded libaudio and fixed CFLAGS, one -ftree-vectorize should be
    enough in Makefile

 -- Bernhard Geier <bernhard.geier@gmx.de>  Thu, 19 Apr 2007 23:38:24 +0200

aacplusenc (0.5) feisty; urgency=low

  * Ported the I/O read/write to fread/fwrite
  * Reduced I/O calls, it encodes much faster now
  * Removed unused au_channel functions
  * Added percent indicator, needed to use aacplusenc as backend

 -- Matteo Croce <rootkit85@yahoo.it>  Mon, 19 Mar 2007 17:40:13 +0100

aacplusenc (0.4) unstable; urgency=low

  * New bitrate tables from 3gpp.org

 -- Matteo Croce <rootkit85@yahoo.it>  Thu, 14 Dec 2006 00:26:01 +0100

mp4tools (0.3) unstable; urgency=low

  * Removed libisomediafile, now aacplusenc saves raw ADTS streams
  * Removed unused downsampling library
  * Now encoding parameters are taken from the WAV header

 -- Matteo Croce <rootkit85@yahoo.it>  Wed,  7 Jun 2006 20:28:35 +0200

mp4tools (0.2) unstable; urgency=low

  * Replaced real's encoder with an opensource, better one

 -- Matteo Croce <rootkit85@yahoo.it>  Tue,  6 Jun 2006 18:44:48 +0200

mp4tools (0.1) unstable; urgency=low

  * Initial release

 -- Matteo Croce <rootkit85@yahoo.it>  Tue, 23 May 2006 15:02:07 +0200


aacplusenc.c
#include <stdio.h>
#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include "FloatFR.h"
#include "au_channel.h"
#include "aacenc.h"
#include "iir32resample.h"
#include "resampler.h"

#include "sbr_main.h"
#include "aac_ram.h"
#include "aac_rom.h"
#include "adts.h"
#include "cfftn.h"
#include "config.h"

/* dynamic buffer of SBR that can be reused for resampling */
extern float sbr_envRBuffer[];

#define CORE_DELAY   (1600)
#define INPUT_DELAY  ((CORE_DELAY)*2 +6*64-2048+1)	/* ((1600 (core codec)*2 (multi rate) + 6*64 (sbr dec delay) - 2048 (sbr enc delay) + magic */
#define MAX_DS_FILTER_DELAY 16	/* the additional max resampler filter delay (source fs) */

#define CORE_INPUT_OFFSET_PS (0)	/* (96-64) makes AAC still some 64 core samples too early wrt SBR ... maybe -32 would be even more correct, but 1024-32 would need additional SBR bitstream delay by one frame */

#define bEncodeMono (inputInfo.nChannels == 1)

/*

      input buffer (1ch)

      |------------ 1537   -------------|-----|---------- 2048 -------------|
           (core2sbr delay     )          ds     (read, core and ds area)
*/
static float
    inputBuffer[(AACENC_BLOCKSIZE * 2 + MAX_DS_FILTER_DELAY +
		 INPUT_DELAY) * MAX_CHANNELS];
char outputBuffer[ADTS_HEADER_SIZE + (6144 / 8) * MAX_CHANNELS];

static IIR21_RESAMPLER IIR21_reSampler[MAX_CHANNELS];

int main(int argc, char *argv[])
{
	AACENC_CONFIG config;

	WavInfo inputInfo;
	FILE *inputFile;

	FILE *hADTSFile;

	int error;

	int bitrate;
	int nChannelsAAC, nChannelsSBR;
	unsigned int sampleRateAAC;
	int frmCnt;
	int bandwidth = 0;

	unsigned int numAncDataBytes = 0;
	unsigned char ancDataBytes[MAX_PAYLOAD_SIZE];
	unsigned int ancDataLength = 0;

	/*!< required only for interfacing with audio output library, thus not counted into RAM requirements */
	short TimeDataPcm[AACENC_BLOCKSIZE * 2 * MAX_CHANNELS];

	int numSamplesRead;
	int bDoIIR2Downsample = 0;
	int bDingleRate = 0;
	int useParametricStereo = 0;
	int coreWriteOffset = 0;
	int coreReadOffset = 0;
	int envWriteOffset = 0;
	int envReadOffset = 0;
	int writeOffset = INPUT_DELAY * MAX_CHANNELS;
	int percent = -1;

	struct AAC_ENCODER *aacEnc = 0;

	int bDoUpsample = 0;
	int upsampleReadOffset = 0;

	int inSamples;
	int bDoIIR32Resample = 0;
	int nSamplesPerChannel;
	const int nRuns = 4;
	float *resamplerScratch = sbr_envRBuffer;

	HANDLE_SBR_ENCODER hEnvEnc = NULL;

	/* initialize the 3GPP instrumenting tool */

	fprintf(stderr,
		"\n"
		"*************************************************************\n"
		"* Enhanced aacPlus Encoder\n"
		"* Build %s, %s\n"
		"* Matteo Croce <rootkit85@yahoo.it>\n"
		"*************************************************************\n\n",
		__DATE__, __TIME__);

	/*
	 * parse command line arguments
	 */
	if (argc != 4) {
		fprintf(stderr,
			"\nUsage:   %s <source.wav> <destination.aac> <bitrate>\n",
			argv[0]);
		fprintf(stderr,
			"\nUse - as filename for stdin and/or stdout.\n");
		fprintf(stderr, "\nExample: %s song.wav song.aac 32\n",
			argv[0]);
		return 0;
	}

	bitrate = atoi(argv[3]) * 1000;;

	fflush(stderr);

	/*
	   open audio input file
	 */

	inputFile = AuChannelOpen(argv[1], &inputInfo);

	if (!inputFile) {
		fprintf(stderr, "could not open %s\n", argv[1]);
		exit(10);
	}

	/*
	   set up basic parameters for aacPlus codec
	 */

	AacInitDefaultConfig(&config);

	nChannelsAAC = nChannelsSBR = inputInfo.nChannels;

	if ((inputInfo.nChannels == 2) && (bitrate >= 16000)
	    && (bitrate < 44001))
		useParametricStereo = 1;

	if (useParametricStereo) {
		nChannelsAAC = 1;
		nChannelsSBR = 2;
	}

	if (((inputInfo.sampleRate == 48000) && (nChannelsAAC == 2)
	     && (bitrate < 24000))
	    || ((inputInfo.sampleRate == 48000) && (nChannelsAAC == 1)
		&& (bitrate < 12000))
	    )
		bDoIIR32Resample = 1;

	if (inputInfo.sampleRate == 16000) {

		bDoUpsample = 1;
		inputInfo.sampleRate = 32000;
		bDingleRate = 1;
	}

	sampleRateAAC = inputInfo.sampleRate;

	if (bDoIIR32Resample)
		sampleRateAAC = 32000;

	config.bitRate = bitrate;
	config.nChannelsIn = inputInfo.nChannels;
	config.nChannelsOut = nChannelsAAC;

	config.bandWidth = bandwidth;

	/*
	   set up SBR configuration
	 */

	if (!IsSbrSettingAvail
	    (bitrate, nChannelsAAC, sampleRateAAC, &sampleRateAAC)) {
		fprintf(stderr,
			"No valid SBR configuration found for:\n\t\t\tbitrate=%d\n\t\t\tchannels=%d\n\t\t\tsamplerate=%d\n",
			bitrate, nChannelsAAC, sampleRateAAC);
		exit(10);
	}

	{
		sbrConfiguration sbrConfig;

		envReadOffset = 0;

		coreWriteOffset = 0;

		if (useParametricStereo) {
			envReadOffset =
			    (MAX_DS_FILTER_DELAY + INPUT_DELAY) * MAX_CHANNELS;
			coreWriteOffset = CORE_INPUT_OFFSET_PS;
			writeOffset = envReadOffset;
		}

		InitializeSbrDefaults(&sbrConfig);

		sbrConfig.usePs = useParametricStereo;

		AdjustSbrSettings(&sbrConfig,
				  bitrate,
				  nChannelsAAC,
				  sampleRateAAC, AACENC_TRANS_FAC, 24000);

		EnvOpen(&hEnvEnc,
			inputBuffer + coreWriteOffset,
			&sbrConfig, &config.bandWidth);

		/* set IIR 2:1 downsampling */

		bDoIIR2Downsample = (bDoUpsample) ? 0 : 1;

		if (useParametricStereo)
			bDoIIR2Downsample = 0;

	}

	/*
	   set up mode-dependant 1:2 upsampling
	 */

	if (bDoUpsample) {

		if (inputInfo.nChannels > 1) {
			fprintf(stderr,
				"\n Stereo @ 16kHz input sample rate is not supported\n");
			return -1;
		}

		InitIIR21_Resampler(IIR21_reSampler);
		InitIIR21_Resampler(IIR21_reSampler + 1);

		assert(IIR21_reSampler[0].delay <= MAX_DS_FILTER_DELAY);

		if (useParametricStereo) {
			writeOffset += AACENC_BLOCKSIZE * MAX_CHANNELS;
			upsampleReadOffset = writeOffset;
			envWriteOffset = envReadOffset;
		} else {
			writeOffset += AACENC_BLOCKSIZE * MAX_CHANNELS;
			coreReadOffset = writeOffset;
			upsampleReadOffset = writeOffset - (((INPUT_DELAY -  IIR21_reSampler[0].delay) >> 1) * MAX_CHANNELS);
			envWriteOffset = ((INPUT_DELAY - IIR21_reSampler[0].delay) & 0x1) * MAX_CHANNELS;
			envReadOffset = 0;
		}
	} else if (bDoIIR2Downsample) {

		/*
		   set up 2:1 downsampling
		 */

			InitIIR21_Resampler(IIR21_reSampler);
			InitIIR21_Resampler(IIR21_reSampler + 1);

			assert(IIR21_reSampler[0].delay <= MAX_DS_FILTER_DELAY);

			writeOffset += IIR21_reSampler[0].delay * MAX_CHANNELS;
		}

	if (bDoIIR32Resample)
		IIR32Init();

	/*
	   set up AAC encoder, now that samling rate is known
	 */

	config.sampleRate = sampleRateAAC;

	error = AacEncOpen(&aacEnc, config);

	if (error) {
		fprintf(stderr, "\n Initialisation of AAC failed !\n");

		AacEncClose(aacEnc);

		AuChannelClose(inputFile);

		return 1;
	}

	if (strcmp(argv[2], "-") == 0)
		hADTSFile = stdout;
	else
		hADTSFile = fopen(argv[2], "wb");

	if (!hADTSFile) {
		fprintf(stderr, "\nFailed to create ADTS file\n");
		exit(10);
	}

	/*
	   Be verbose
	 */

	fprintf(stderr, "input file %s: \nsr = %d, nc = %d\n\n", argv[1],
		inputInfo.sampleRate, inputInfo.nChannels);
	fprintf(stderr, "output file %s: \nbr = %d sr-OUT = %d  nc-OUT = %d\n\n",
		argv[2], bitrate, sampleRateAAC * 2, nChannelsSBR);
	fflush(stderr);

	init_plans();

	frmCnt = 0;

	memset(TimeDataPcm, 0, sizeof(TimeDataPcm));

	/*
	   set up input samples block size feed
	 */

	if (bDoIIR32Resample) {
		inSamples = IIR32GetResamplerFeed(AACENC_BLOCKSIZE *  inputInfo.nChannels * 2) / nRuns;
		assert(inSamples <= AACENC_BLOCKSIZE * 2);
	} else {
		inSamples = AACENC_BLOCKSIZE * inputInfo.nChannels * 2;
		if (bDoUpsample)
			inSamples = inSamples >> 1;
	}

	/* create the ADTS header */
	adts_hdr(outputBuffer, &config);

	/*
	   The frame loop
	 */
	while (1) {
		int i, ch, outSamples, numOutBytes, newpercent = -1;

		/*
		   File input read, resample and downmix
		 */

		if (bDoIIR32Resample) {

			/* resampling from 48 kHz to 32 kHz prior to encoding */

			int stopLoop = 0;
			int nSamplesProcessed = 0;
			int r;

			const int nDSOutBlockSize = AACENC_BLOCKSIZE * 2 / nRuns;
			int stride = inputInfo.nChannels;
			if (inputInfo.nChannels == 1)
				stride = 2;

			/* counting previous operations */

			for (r = 0; r < nRuns; r++) {
				if (AuChannelReadShort(inputFile, TimeDataPcm,  inSamples, &numSamplesRead)) {
					stopLoop = 1;
					break;
				}

				/* copy from short to float input buffer */
				/* resamplerScratch[]
				   TimeDataPcm[]
				 */
				for (i = 0; i < numSamplesRead; i++)
					resamplerScratch[i] = (float)TimeDataPcm[i];

				switch (inputInfo.nChannels) {
				case 1:
					nSamplesPerChannel = numSamplesRead;
					break;
				case 2:
					nSamplesPerChannel = numSamplesRead >> 1;
					break;
				default:
					nSamplesPerChannel = numSamplesRead / inputInfo.nChannels;
				}

				/* supposed to return exact number samples to encode one block of audio */

				nSamplesProcessed +=
				    IIR32Resample(resamplerScratch,
						  &inputBuffer[writeOffset +  r * nDSOutBlockSize * stride],
						  nSamplesPerChannel,
						  nDSOutBlockSize,
						  inputInfo.nChannels);

				if (inputInfo.nChannels == 1 && stride == 2)
					/* reordering necessary since the encoder takes interleaved data */
					for (i = nDSOutBlockSize - 1; i >= 0;  i--)
						inputBuffer[writeOffset + r * nDSOutBlockSize * 2 + 2 * i] =
						    inputBuffer[writeOffset + r * nDSOutBlockSize * 2 + i];
			}

			if (stopLoop)
				break;

			numSamplesRead = nSamplesProcessed;

		} else {
			/* no resampling prior to encoding required */
			/* read from file */
			if (AuChannelReadShort(inputFile, TimeDataPcm,
					       inSamples, &numSamplesRead))
				break;

			/* copy from short to float input buffer */

			/* inputBuffer[i+writeOffset]
			   TimeDataPcm[i]
			 */
			if (inputInfo.nChannels == nChannelsSBR)
				for (i = 0; i < numSamplesRead; i++)
					inputBuffer[i + writeOffset] = (float)TimeDataPcm[i];

			/* copy from short to float input buffer, reordering necessary since the encoder takes interleaved data */
			if (inputInfo.nChannels == 1)
				/* inputBuffer[writeOffset+MAX_CHANNELS*i]
				   inputBuffer[writeOffset+i]
				 */
				for (i = 0; i < numSamplesRead; i++)
					inputBuffer[writeOffset + 2 * i] = (float)TimeDataPcm[i];

			/* copy from short to float input buffer, downmix stereo input signal to mono, reordering necessary since the encoder takes interleaved data */

			if ((inputInfo.nChannels == 2) && bEncodeMono)
				/*  inputBuffer[writeOffset+2*i]
				   inputBuffer[writeOffset+2*i+1]
				 */
				for (i = 0; i < numSamplesRead / 2; i++)
					inputBuffer[writeOffset + 2 * i] = ((float)TimeDataPcm[2 * i] + (float)TimeDataPcm[2 * i + 1]) * 0.5f;
		}		/*end if (bDoIIR32Resample) */

		if (bDoUpsample)
			for (ch = 0; ch < inputInfo.nChannels; ch++)
				IIR21_Upsample(&(IIR21_reSampler[ch]),
					       inputBuffer +
					       upsampleReadOffset + ch,
					       numSamplesRead /
					       inputInfo.nChannels,
					       MAX_CHANNELS,
					       inputBuffer + envWriteOffset +
					       ch, &outSamples, MAX_CHANNELS);

		/*
		   encode one SBR frame
		 */
		EnvEncodeFrame(hEnvEnc,
			       inputBuffer + envReadOffset,
			       inputBuffer + coreWriteOffset,
			       MAX_CHANNELS, &numAncDataBytes, ancDataBytes);

		/*
		   2:1 downsampling for AAC core
		 */
		if (bDoIIR2Downsample)
			for (ch = 0; ch < nChannelsAAC; ch++)
				IIR21_Downsample(&(IIR21_reSampler[ch]),
						 inputBuffer + writeOffset + ch,
						 numSamplesRead /
						 inputInfo.nChannels,
						 MAX_CHANNELS, inputBuffer + ch,
						 &outSamples, MAX_CHANNELS);

		/* SBR side info data is passed as ancillary data */
		if (numAncDataBytes == 0)
			numAncDataBytes = ancDataLength;

		/*
		   encode one AAC frame
		 */
		if (hEnvEnc && useParametricStereo) {
			AacEncEncode(aacEnc, inputBuffer, 1,	/* stride */
				     ancDataBytes,
				     &numAncDataBytes,
				     (unsigned *)(outputBuffer + ADTS_HEADER_SIZE),
				     &numOutBytes);

			if (hEnvEnc)
				memcpy(inputBuffer,
				       inputBuffer + AACENC_BLOCKSIZE,
				       CORE_INPUT_OFFSET_PS * sizeof(float));
		} else {
			AacEncEncode(aacEnc,
				     inputBuffer + coreReadOffset,
				     MAX_CHANNELS,
				     ancDataBytes,
				     &numAncDataBytes,
				     (unsigned *)(outputBuffer + ADTS_HEADER_SIZE),
				     &numOutBytes);

			if (hEnvEnc) {
				if (bDoUpsample) {
					memmove(&inputBuffer[envReadOffset],
						&inputBuffer[envReadOffset +  AACENC_BLOCKSIZE * MAX_CHANNELS * 2],
						(envWriteOffset - envReadOffset) *
						sizeof(float));

					memmove(&inputBuffer
						[upsampleReadOffset],
						&inputBuffer[upsampleReadOffset + AACENC_BLOCKSIZE * MAX_CHANNELS],
						(writeOffset - upsampleReadOffset) *
						sizeof(float));
				} else
					memmove(inputBuffer,
						inputBuffer +
						AACENC_BLOCKSIZE * 2 *
						MAX_CHANNELS,
						writeOffset * sizeof(float));
			}
		}

		/*
		   Write one frame of encoded audio to file
		 */

		if (numOutBytes) {
			adts_hdr_up(outputBuffer, numOutBytes);
			fwrite(outputBuffer, 1, numOutBytes + ADTS_HEADER_SIZE, hADTSFile);
		}

		frmCnt++;

		/* 3GPP instrumenting tool: measure worst case work load at end of each decoding loop */

		newpercent = frmCnt * AACENC_BLOCKSIZE * inputInfo.nChannels / (inputInfo.nSamples / (4 * 100));
		if (newpercent != percent) {
			percent = newpercent;
			fprintf(stderr, "[%d%%]\r", newpercent);
			fflush(stderr);
		}
	}
	fprintf(stderr, "\n");
	fflush(stderr);

	/*
	   Close encoder
	 */

	AacEncClose(aacEnc);

	AuChannelClose(inputFile);

	destroy_plans();

	if (hEnvEnc)
		EnvClose(hEnvEnc);

	/* 3GPP instrumenting tool: print output data */

	fprintf(stderr, "\nencoding finished\n");
	fflush(stderr);

	return 0;
}
